File: ecos-records/src/main/java/ru/citeck/ecos/records3/rest/RestHandlerAdapter.java
package ru.citeck.ecos.records3.rest;

import ecos.com.fasterxml.jackson210.databind.JsonNode;
import ecos.com.fasterxml.jackson210.databind.node.ObjectNode;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import ru.citeck.ecos.commons.json.Json;
import ru.citeck.ecos.commons.json.JsonMapper;
import ru.citeck.ecos.records2.RecordsServiceFactory;
import ru.citeck.ecos.records2.request.rest.DeletionBody;
import ru.citeck.ecos.records2.request.rest.MutationBody;
import ru.citeck.ecos.records2.request.rest.RestHandler;
import ru.citeck.ecos.records3.rest.v1.delete.DeleteBody;
import ru.citeck.ecos.records3.rest.v1.mutate.MutateBody;
import ru.citeck.ecos.records3.rest.v1.query.QueryBody;
import ru.citeck.ecos.records3.rest.v1.RestHandlerV1;

@Slf4j
public class RestHandlerAdapter {

    private final RestHandler restHandlerV0;
    private final RestHandlerV1 restHandlerV1;
    private final JsonMapper mapper = Json.getMapper();

    public RestHandlerAdapter(RecordsServiceFactory serviceFactory) {
        restHandlerV1 = new RestHandlerV1(serviceFactory);
        restHandlerV0 = serviceFactory.getRestHandler();
    }

    public Object queryRecords(Object body) {

        BodyWithVersion bodyWithVersion = getBodyWithVersion(body);

        switch (bodyWithVersion.version) {
            case 0:

                ru.citeck.ecos.records2.request.rest.QueryBody v0Body =
                    mapper.convert(bodyWithVersion.getBody(),
                        ru.citeck.ecos.records2.request.rest.QueryBody.class);

                if (v0Body == null) {
                    v0Body = new ru.citeck.ecos.records2.request.rest.QueryBody();
                }
                return restHandlerV0.queryRecords(v0Body);

            case 1:

                return restHandlerV1.queryRecords(mapper.convert(bodyWithVersion, QueryBody.class));

            default:
                throw new IllegalArgumentException("Unknown body version. Body: " + bodyWithVersion);
        }
    }

    public Object deleteRecords(Object body) {

        BodyWithVersion bodyWithVersion = getBodyWithVersion(body);

        switch (bodyWithVersion.version) {
            case 0:

                DeletionBody v0Body = mapper.convert(bodyWithVersion.getBody(), DeletionBody.class);

                if (v0Body == null) {
                    v0Body = new DeletionBody();
                }
                return restHandlerV0.deleteRecords(v0Body);

            case 1:

                return restHandlerV1.deleteRecords(mapper.convert(bodyWithVersion, DeleteBody.class));

            default:
                throw new IllegalArgumentException("Unknown body version. Body: " + bodyWithVersion);
        }
    }

    public Object mutateRecords(Object body) {

        BodyWithVersion bodyWithVersion = getBodyWithVersion(body);

        switch (bodyWithVersion.version) {
            case 0:

                MutationBody v0Body = mapper.convert(bodyWithVersion.getBody(), MutationBody.class);

                if (v0Body == null) {
                    v0Body = new MutationBody();
                }
                return restHandlerV0.mutateRecords(v0Body);

            case 1:

                return restHandlerV1.mutateRecords(mapper.convert(bodyWithVersion, MutateBody.class));

            default:
                throw new IllegalArgumentException("Unknown body version. Body: " + bodyWithVersion);
        }
    }

    private BodyWithVersion getBodyWithVersion(Object body) {

        ObjectNode jsonBody = Json.getMapper().convert(body, ObjectNode.class);
        if (jsonBody == null) {
            throw new IllegalArgumentException("Incorrect request body. Expected JSON Object, but found: " + body);
        }

        JsonNode version = jsonBody.path("version");
        if (version.isNumber()) {
            return new BodyWithVersion(jsonBody, version.asInt());
        }

        JsonNode v1Body = jsonBody.path("v1Body");
        if (v1Body instanceof ObjectNode && v1Body.size() > 0) {
            return new BodyWithVersion((ObjectNode) v1Body, 1);
        }

        return new BodyWithVersion(jsonBody, 0);
    }

    @Data
    @AllArgsConstructor
    private static class BodyWithVersion {
        private final ObjectNode body;
        private final int version;
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/rest/RestHandlerAdapter.kt
package ru.citeck.ecos.records3.rest

import com.fasterxml.jackson.databind.node.ObjectNode
import ru.citeck.ecos.commons.json.Json
import ru.citeck.ecos.records3.RecordsServiceFactory
import ru.citeck.ecos.records3.record.resolver.RemoteRecordsResolver
import ru.citeck.ecos.records3.rest.v1.RestHandlerV1
import ru.citeck.ecos.records3.rest.v1.delete.DeleteBody
import ru.citeck.ecos.records3.rest.v1.mutate.MutateBody
import ru.citeck.ecos.records3.rest.v2.query.QueryBodyV2
import ru.citeck.ecos.webapp.api.web.executor.EcosWebExecutor
import ru.citeck.ecos.webapp.api.web.executor.EcosWebExecutorReq
import ru.citeck.ecos.webapp.api.web.executor.EcosWebExecutorResp
import ru.citeck.ecos.webapp.api.web.executor.EcosWebExecutorsApi
import ru.citeck.ecos.records3.rest.v1.query.QueryBody as QueryBodyV1

class RestHandlerAdapter(services: RecordsServiceFactory) {

    companion object {
        const val UNKNOWN_BODY_VERSION_MSG = "Unknown body version"
    }

    private val restHandlerV1 = RestHandlerV1(services)
    private val mapper = Json.mapper

    init {
        registerWebExecutors(services.getEcosWebAppApi()?.getWebExecutorsApi())
    }

    private fun registerWebExecutors(executors: EcosWebExecutorsApi?) {
        executors ?: return
        executors.register(
            object : EcosWebExecutor {
                override fun execute(request: EcosWebExecutorReq, response: EcosWebExecutorResp) {
                    val result = queryRecords(
                        request.getBodyReader().readDto(ObjectNode::class.java),
                        request.getApiVersion()
                    )
                    response.getBodyWriter().writeDto(result)
                }

                // Note: API version 0 is unsupported. However, the EcosLegacyWebClient
                // does not utilize the WebAPI if version 0 is not provided here.
                // This is a temporary solution until all microservices are upgraded
                // to Spring Boot 3+ and Java 21+.
                override fun getApiVersion() = 0 to 2
                override fun getPath() = RemoteRecordsResolver.QUERY_PATH
                override fun isReadOnly(): Boolean = true
            }
        )
        executors.register(
            object : EcosWebExecutor {
                override fun execute(request: EcosWebExecutorReq, response: EcosWebExecutorResp) {
                    val result = mutateRecords(
                        request.getBodyReader().readDto(ObjectNode::class.java),
                        request.getApiVersion()
                    )
                    response.getBodyWriter().writeDto(result)
                }

                // Note: API version 0 is unsupported. However, the EcosLegacyWebClient
                // does not utilize the WebAPI if version 0 is not provided here.
                // This is a temporary solution until all microservices are upgraded
                // to Spring Boot 3+ and Java 21+.
                override fun getApiVersion() = 0 to 1
                override fun getPath() = RemoteRecordsResolver.MUTATE_PATH
                override fun isReadOnly(): Boolean = false
            }
        )
        executors.register(
            object : EcosWebExecutor {
                override fun execute(request: EcosWebExecutorReq, response: EcosWebExecutorResp) {
                    val result = deleteRecords(
                        request.getBodyReader().readDto(ObjectNode::class.java),
                        request.getApiVersion()
                    )
                    response.getBodyWriter().writeDto(result)
                }

                // Note: API version 0 is unsupported. However, the EcosLegacyWebClient
                // does not utilize the WebAPI if version 0 is not provided here.
                // This is a temporary solution until all microservices are upgraded
                // to Spring Boot 3+ and Java 21+.
                override fun getApiVersion() = 0 to 1
                override fun getPath() = RemoteRecordsResolver.DELETE_PATH
                override fun isReadOnly(): Boolean = false
            }
        )
    }

    fun queryRecords(body: ObjectNode, version: Int): Any {

        return when (version) {
            1, 2 -> {
                var bodyData = body
                val queryType: Class<out QueryBodyV1> = if (version == 2) {
                    QueryBodyV2::class.java
                } else {
                    val afterId = bodyData.path("query").path("page").path("afterId")
                    if (afterId.isTextual && afterId.asText().isBlank()) {
                        bodyData = bodyData.deepCopy()
                        (bodyData.path("query").path("page") as? ObjectNode)?.remove("afterId")
                    }
                    QueryBodyV1::class.java
                }
                val v1Body = mapper.convert(bodyData, queryType) ?: QueryBodyV1()
                restHandlerV1.queryRecords(v1Body)
            }
            else -> error("$UNKNOWN_BODY_VERSION_MSG: $version")
        }
    }

    fun deleteRecords(body: ObjectNode, version: Int): Any {

        return when (version) {
            1 -> {
                val v1Body = mapper.convert(body, DeleteBody::class.java) ?: DeleteBody()
                restHandlerV1.deleteRecords(v1Body)
            }
            else -> error("$UNKNOWN_BODY_VERSION_MSG: $version")
        }
    }

    fun mutateRecords(body: ObjectNode, version: Int): Any {

        return when (version) {
            1 -> {
                val v1Body = mapper.convert(body, MutateBody::class.java) ?: MutateBody()
                restHandlerV1.mutateRecords(v1Body)
            }
            else -> error("$UNKNOWN_BODY_VERSION_MSG: $version")
        }
    }

    fun getV1Handler(): RestHandlerV1 {
        return restHandlerV1
    }
}
